go-proio: Improved Reader.Events() scan

decibelcooper opened the pull request:
As pointed out by @sbinet, Reader.Events() allowed for leaking
goroutines if the channel was not exhausted.  This has been resolved by
adding the ability to stop the scan and close the channel.

Reader.Events() has been renamed to Reader.ScanEvents(), and another
function Reader.StopScan() has been added.  Additionally, Reader.Close()
will call Reader.StopScan().  A mutex for entering Reader.Get()-like
functions has been added for thread safety and to allow multiple
scanners to exist simultaneously.

Finally, the channel has been made buffered, and the variable
Reader.EventScanBufferSize has been added with a default value of 100.
This can be used to customize the depth of the buffered channel.

This PR resolves #6

decibelcooper committed:
go-proio: Improved Reader.Events() scan

As pointed out by @sbinet, Reader.Events() allowed for leaking
goroutines if the channel was not exhausted.  This has been resolved by
adding the ability to stop the scan and close the channel.

Reader.Events() has been renamed to Reader.ScanEvents(), and another
function Reader.StopScan() has been added.  Additionally, Reader.Close()
will call Reader.StopScan().  A mutex for entering Reader.Get()-like
functions has been added for thread safety and to allow multiple
scanners to exist simultaneously.

Finally, the channel has been made buffered, and the variable
Reader.EventScanBufferSize has been added with a default value of 100.
This can be used to customize the depth of the buffered channel.

decibelcooper commented:
Still need to resolve how to handle errors with buffered channel.  Currently, Reader.Err just gives the most recent error, which does not correspond to the current event in general since the channel is buffered.

sbinet commented on go-proio/reader.go:
```diff
@@ -120,33 +127,41 @@ var (
 // event, and ErrResync will be returned along with the event.
 func (rdr *Reader) Get() (*Event, error) {
+	rdr.getMutex.Lock()
+
```
you might as well use `defer rdr.getMutex.Unlock()` here instead of sprinkling it everywhere :)

decibelcooper replied:
I actually was just about to say that too, thanks :)

sbinet commented on go-proio/reader.go:
```diff
@@ -159,60 +174,98 @@ func (rdr *Reader) Get() (*Event, error) {
 	}()
+
+	stopThisScan := func() {
+		scanLoop = false
```
that's a race, I believe.

the canonical way to stop such a goroutine would be to:

```go
events := make(chan *Event, rdr.EventScanBufferSize)
quit := make(chan int)

go func() {
    defer close(events)
    for {
        evt, _ := rdr.Get()
        if evt == nil { break }
        select {
        case events <- evt:
        case <-quit:
            return
        }
    }
}()

stopThisScan := func () { close(quit) }()
```

decibelcooper replied:
You're right, there is a possible race condition.  I should do it this way instead.

sbinet commented on go-proio/reader.go:
```diff
@@ -159,60 +174,98 @@ func (rdr *Reader) Get() (*Event, error) {
 // stream.
 func (rdr *Reader) GetHeader() (*model.EventHeader, error) {
+	rdr.getMutex.Lock()
+
```
here too, you could use a `defer rdr.getMutex.Unlock()`

sbinet commented on go-proio/reader.go:
```diff
@@ -223,18 +276,22 @@ func (rdr *Reader) GetHeader() (*model.EventHeader, error) {
 // Skip the next nEvents events
 func (rdr *Reader) Skip(nEvents int) (int, error) {
+	rdr.getMutex.Lock()
+
```
here as well.

sbinet commented on go-proio/reader.go:
```diff
@@ -280,20 +342,26 @@ var ErrNotSeekable = errors.New("data stream is not seekable")
 // beginning of the file.
 func (rdr *Reader) SeekToStart() error {
+	rdr.getMutex.Lock()
+
```
_ditto_

decibelcooper committed:
go-proio: Changed error handling with Reader.ScanEvents()

Errors are now pushed to the Reader.Err channel when Reader.ScanEvents()
is called.
